feat(language-service): Allow auto-imports of a pipe via quick fix when its selector is used.

dylhunn opened the pull request:
Four commits, as below:

### refactor(language-service): Improve the quick fix auto-import tests.

Currently, the auto-import tests are very difficult to debug, because the `expect`ations are buried several function calls deep, and also because the multi-line replacements are hard to deal with. The auto-import tests now use a different approach, which can be easily debugged from the failure message. In addition, several new tests have been added to cover more cases.

### refactor(compiler-cli): Make `getKnown` return an array of nodes.

`MetadataReaderWithIndex.getKnown` currently returns an iterator. It will be easier to work with for upcoming usages if it returns an array instead.

### feat(language-service): Introduce a new NgModuleIndex, and use it to suggest re-exports.

NgModules can re-export other NgModules, which transitively includes all traits exported by the re-exported NgModule. We want to be able to suggest *all* re-exports of a component when trying to auto-import it.

Previously, we used an approximation when importing from NgModules: we looked at a Component's metadata, and just imported the declaring NgModule. However, this is not technically correct -- the declaring NgModule it is not necessarily the same one that exports it for the current scope. (Indeed, there could be multiple re-exports!) As a replacement, I have implemented a more general solution.

This PR introduces a new class on the template type checker, called `NgModuleIndex`. When queried, it conducts a depth-first-search over the NgModule import/export graph, in order to find all NgModules anywhere in the current dependency graph, as well as all exports of those NgModules. This allows the language service to suggest all of the re-exports, in addition to the original export.

### feat(language-service): Allow auto-imports of a pipe via quick fix when its selector is used, both directly and via reexports.

A previous PR introduced a new compiler abstraction that tracks *all* known exports and re-exports of Angular traits. This PR harnesses that abstraction in the language service, in order to allow automatic imports of pipes.

dylhunn committed:
refactor(language-service): Improve the quick fix auto-import tests.

Currently, the auto-import tests are very difficult to debug, because the `expect`ations are buried several function calls deep, and also because the multi-line replacements are hard to deal with. The auto-import tests now use a different approach, which can be easily debugged from the failure message. In addition, several new tests have been added to cover more cases.

atscott commented on packages/language-service/src/codefixes/fix_missing_import.ts:
```diff
@@ -40,43 +42,42 @@ function getCodeActions(
   }
-  const missingElement = target.context.node;
 
+  let matches: Set<PotentialDirective|PotentialPipe> = new Set();
```
should this be `Set<PotentialDirective>|Set<PotentialPipe>` instead?

atscott reviewed (approved):
reviewed-for: fw-language-service

atscott reviewed (approved):
reviewed-for: language-service

JoostK commented on packages/compiler-cli/src/ngtsc/metadata/src/global.ts:
```diff
@@ -0,0 +1,156 @@
+      }
+      let explore = index.all();
+      explore = explore.filter(d => visited.has(d.ref.node));
+      explore.forEach(d => visited.add(d.ref.node));
```
We tend to not use `forEach` in the compiler code, in favor of `for..of` loops.

JoostK commented on packages/compiler-cli/src/ngtsc/metadata/src/global.ts:
```diff
@@ -0,0 +1,156 @@
+  private ensureIndex() {
+    if (this.indexed) return;
+    this.indexed = true;
+    this.metaForTrait = {0: new Map(), 1: new Map(), 2: new Map()};
```
Additionally, I am confused by why this is lazily initialized versus the other fields which are not. By not initializing it you don't really need the `indexed` boolean state, as it can be derived from `metaForTrait` being uninitialized.

dylhunn replied:
Fixed by eagerly initializing the map -- I do like having the clear boolean indicator

JoostK commented on packages/compiler-cli/src/ngtsc/metadata/src/global.ts:
```diff
@@ -0,0 +1,156 @@
+  private indexed = false;
+  private indexFor = new Map<ClassDeclaration, ImportExportIndex>();
+  private ngModulesExportingTrait = new Map<ClassDeclaration, Set<NgModuleMeta>>();
+  private metaForTrait?: {[K in MetaKind]: Map<ClassDeclaration, MetaByKind<K>>};
```
We tend to avoid optional properties in favor of explicit null initializers.

JoostK commented on packages/compiler-cli/src/ngtsc/metadata/src/global.ts:
```diff
@@ -0,0 +1,156 @@
+      const trait = stack.pop()!;
+      const index =
+          cache(this.indexFor, trait.ref.node, () => new ImportExportIndex(trait, this.metaReader));
+      cache(this.metaForTrait[trait.kind], trait.ref.node, () => trait);
```
This seems really strange to me. An explicit assignment would do the same:

```suggestion
      this.metaForTrait[trait.kind].set(trait.ref.node, trait);
```

JoostK commented on packages/compiler-cli/src/ngtsc/typecheck/src/checker.ts:
```diff
@@ -687,46 +688,59 @@ export class TemplateTypeCheckerImpl implements TemplateTypeChecker {
 
-    // Import the ngModule if one exists. Otherwise, import the standalone trait directly.
-    const importTarget = ngModuleRef ?? toImport.ref;
+    let imports = new Array<PotentialImport>();
```
`let` => `const`

JoostK commented on packages/compiler-cli/src/ngtsc/typecheck/testing/index.ts:
```diff
@@ -577,17 +578,52 @@ function getFakeMetadataReader(fakeMetadataRegistry: Map<any, DirectiveMeta|null
           return fakeMetadataRegistry.get(node.debugName) ?? null;
         },
-    getKnown(kind: MetaKind): Iterable<ClassDeclaration> {
+    getKnown(kind: MetaKind): Array<ClassDeclaration> {
```
style nit:

```suggestion
    getKnown(kind: MetaKind): ClassDeclaration[] {
```

JoostK commented on packages/compiler-cli/src/ngtsc/typecheck/testing/index.ts:
```diff
@@ -577,17 +578,52 @@ function getFakeMetadataReader(fakeMetadataRegistry: Map<any, DirectiveMeta|null
+function getFakeGlobalReader(fakeMetadataRegistry: Map<any, DirectiveMeta|null>):
+    MetadataReaderWithIndex&NgModuleIndex {
+  return {
+    getDirectiveMetadata(node: Reference<ClassDeclaration>): DirectiveMeta |
+        null {
+          return fakeMetadataRegistry.get(node.debugName) ?? null;
+        },
+    getPipeMetadata(node: Reference<ClassDeclaration>): PipeMeta |
+        null {
+          return null;
+        },
+    getNgModuleMetadata(node: Reference<ClassDeclaration>): NgModuleMeta |
+        null {
+          return null;
+        },
```
clang-format 😖

dylhunn replied:
Agreed, but I'm not usually inclined to turn it off unless it's totally unreadable, so 🤷

JoostK commented on packages/language-service/src/codefixes/fix_missing_import.ts:
```diff
@@ -40,43 +42,42 @@ function getCodeActions(
-
-  for (let currMatch of matches) {
-    const currMatchSymbol = currMatch.tsSymbol.valueDeclaration;
+  for (let currMatch of matches.values()) {
```
```suggestion
  for (const currMatch of matches.values()) {
```

JoostK commented on packages/language-service/src/codefixes/fix_missing_import.ts:
```diff
@@ -40,43 +42,42 @@ function getCodeActions(
   }
-  const missingElement = target.context.node;
 
+  let matches: Set<PotentialDirective>|Set<PotentialPipe> = new Set<PotentialDirective>();
```
If `matches` stays empty then this method returns an empty array, so I'd suggest removing the initializer expression here and adding an `else { return []; }` block to the below if-statement.

JoostK commented on packages/language-service/src/codefixes/fix_missing_import.ts:
```diff
@@ -40,43 +42,42 @@ function getCodeActions(
+  } else if (
+      target.context.kind === TargetNodeKind.RawExpression &&
+      target.context.node instanceof ASTWithName) {
+    const name = (target.context.node as any).name;
```
The `any` cast seems suspicious here, given that this should have been narrowed to be of type `ASTWithName`?

dylhunn replied:
Indeed, ASTWithName doesn't seem to actually declare the `name` key (even though it's there, and that's the point!) -- I will get to the bottom of this.

JoostK reviewed (commented):
Initial pass with mostly stylistic comments; I haven't looked too deeply in how this all works,

dylhunn commented on packages/compiler-cli/src/ngtsc/metadata/src/global.ts:
```diff
@@ -0,0 +1,156 @@
+    }
+    if (this._imports === undefined) {
+      this._imports = {
+        0: this.meta.imports?.map((m) => this.reader.getDirectiveMetadata(m)).filter(nonNull) ?? [],
```
Yes, thanks! I didn't realize you could put enumerated keys in square brackets like this :)

dylhunn commented:
@JoostK Thank you for your first review pass!

alxhub commented on packages/compiler-cli/src/ngtsc/metadata/src/global.ts:
```diff
@@ -0,0 +1,166 @@
+ * For a particular metadata object (DirectiveMeta, PipeMeta, or NgModuleMeta), keeps track of all
+ * the imports and exports of various kinds from that metadata object.
+ */
+class ImportExportIndex {
```
After reading through this class a few times, I'm starting to wonder if it makes sense as an abstraction. We already have the `GlobalMetadataReader` which populates a cache of information across the whole program. The operations going on inside `ImportExportIndex` should probably be done there, and the results stored in the cache as plain objects instead of this smart data object / hybrid class.

alxhub commented on packages/compiler-cli/src/ngtsc/metadata/src/global.ts:
```diff
@@ -0,0 +1,166 @@
+    const localNgModules =
+        this.localMetaReader.getKnown(MetaKind.NgModule)
+            .map(m => this.localMetaReader.getNgModuleMetadata(new Reference(m)));
+    const stack = [...localDirectives, ...localPipes, ...localNgModules].filter(nonNull);
```
`stack` should probably track just the `ClassDeclaration`, and we can process each trait separately.

alxhub commented on packages/compiler-cli/src/ngtsc/metadata/src/global.ts:
```diff
@@ -0,0 +1,166 @@
+    return Array.from(this.metaForTrait![kind].keys());
+  }
+
+  getDirectiveMetadata(ref: Reference<ClassDeclaration>): DirectiveMeta|null {
```
This seems like it's duplicating work that the existing metadata reader system can already do. Do we need to support these operations here as well?

alxhub commented on packages/compiler-cli/src/ngtsc/metadata/src/global.ts:
```diff
@@ -0,0 +1,166 @@
+    }
+  }
+
+  getNgModulesExporting(trait: ClassDeclaration): NgModuleMeta[]|null {
```
We could consider returning `ClassDeclaration[]` or `Reference<ClassDeclaration>[]` here instead of `NgModuleMeta`. That would reduce the responsibilities of this class to just answering a simple question ("who exports this thing") and make it less of a swiss army knife of metadata operations.

alxhub commented on packages/compiler-cli/src/ngtsc/metadata/src/global.ts:
```diff
@@ -0,0 +1,166 @@
+ * includes types defined in the current compilation unit, as well as all upstream transitive
+ * exports and re-exports.
+ */
+export class GlobalMetadataReader implements MetadataReaderWithIndex, NgModuleIndex {
```
This is an interesting idea, but I think also a bit overengineered for the problem. The normal `MetadataReader` used by the compiler _can_ give you metadata about any type you ask about. The real operation that `GlobalMetadataReader` here is doing is walking the NgModule graph to understand dependencies - that's a very different operation than "reading metadata" about any individual type.

alxhub reviewed (changes requested):
Some initial comments on the `GlobalMetadataReader` implementation.

Also, we should separate out that implementation from the usage in auto-imports as a separate commit. Ideally we'd have 3 commits:

* One commit that Implements the new service type (currently `GlobalMetadataReader` but as mentioned I think we need a different concept)
* One commit that ports the existing quick fix over to using it instead of `getOwningNgModule`
* A final `feat()` commit that implements auto-import of pipes.

dylhunn commented:
Thanks. I'm going to work on fixing these design issues and I'll comment again when it's ready.

dylhunn committed:
feat(language-service): Allow auto-imports of a pipe via quick fix when its selector is used, both directly and via reexports.

A previous PR introduced a new compiler abstraction that tracks *all* known exports and re-exports of Angular traits. This PR harnesses that abstraction in the language service, in order to allow automatic imports of pipes.

dylhunn commented:
PR has been updated with a new design.

dylhunn commented:
Hi @alxhub! Please let me know if this looks OK to you when you have some time, thanks!

dylhunn commented:
@alxhub Please let me know if you'd had a chance to look at this, thanks! I'm blocking @crisbeto so I'd like to get all the issues fixed ASAP, preferably this week.

JoostK commented on packages/compiler-cli/src/ngtsc/typecheck/src/checker.ts:
```diff
@@ -687,46 +688,59 @@ export class TemplateTypeCheckerImpl implements TemplateTypeChecker {
 
-    // Import the ngModule if one exists. Otherwise, import the standalone trait directly.
-    const importTarget = ngModuleRef ?? toImport.ref;
+    const imports = new Array<PotentialImport>();
```
```suggestion
    const imports: PotentialImport[] = [];
```

JoostK commented on packages/compiler-cli/src/ngtsc/typecheck/src/checker.ts:
```diff
@@ -687,46 +688,59 @@ export class TemplateTypeCheckerImpl implements TemplateTypeChecker {
-      return [];
+
+    const exportingNgModules = this.ngModuleIndex.getNgModulesExporting(meta);
+    if (exportingNgModules === null) {
+      return imports;
```
nit: I'd invert this condition and avoid the early-return here. This method gathers the imports from multiple sources and it is a bit error-prone to allow the last source to exit the method halfway; any additional sources now cannot simply be appended, as they'd then be affected by this early-return that is specific for a single source.

JoostK commented on packages/compiler-cli/src/ngtsc/metadata/src/ng_module_index.ts:
```diff
@@ -0,0 +1,100 @@
+      from: Reference<ClassDeclaration>, target: ClassDeclaration, results: Set<ClassDeclaration>,
+      references: Map<ClassDeclaration, Set<Reference<ClassDeclaration>>>,
+      visited: Set<ClassDeclaration>): boolean {
+    if (from.node === target) {
+      return true;
+    }
```
This condition seems a bit strange here; `from.node === target` implies that `from` is a directive, but directives don't export themselves so returning `true` for the question `doesExport` is confusing.

I think the confusion can be avoided by pulling up this condition into the `fromNgModule.exports` loop below. That would most likely allow this method to drop the `boolean` return type entirely which simplifies things even more, and then allows for a more appropriate method name (the current calls to a method named `doesExport` where the return value is not used currently is really confusing.)

dylhunn replied:
Indeed. I did this to avoid having an arm's-length base case (e.g. inside a condition), but your way probably is clearer.

dylhunn replied:
(Note however that we cannot remove the boolean return type, because we need it in order for transitive re-exports to be properly added further up the call stack.)

JoostK commented on packages/compiler-cli/src/ngtsc/metadata/src/ng_module_index.ts:
```diff
@@ -0,0 +1,100 @@
+    for (const dirOrNgModule of locallyDeclaredDirsAndNgModules) {
+      this.doesExport(dirOrNgModule, directiveOrPipe.ref.node, results, references, visited);
+    }
+    const referencesForResults =
+        Array.from(results).map(r => references.get(r) ?? null).filter(nonNull);
+    const resultMetas = referencesForResults.flatMap(
+        refSet => Array.from(refSet)
+                      .map(ref => this.metaReader.getNgModuleMetadata(ref))
+                      .filter(nonNull));
+    const deduped = [...new Set(resultMetas)];
```
Call me old-fashioned but a `for-of` loop seems a lot more straightforward here:

```ts
const metas: NgModuleMetadata[] = [];
for (const result of results) {
  const refs = references.get(result);
  if (refs === null) {
    continue;
  }

  for (const ref of refs) {
    const meta = this.metaReader.getNgModuleMetadata(ref);
    if (meta !== null) {
      metas.push(meta);
    }
  }
}
this.resultCache.set(directiveOrPipe.ref.node, Array.from(metas));
```

dylhunn replied:
Sure. I do tend to use `map`/`filter` more than most :)

JoostK commented on packages/compiler-cli/src/ngtsc/metadata/src/ng_module_index.ts:
```diff
@@ -0,0 +1,100 @@
+    }
+    const locallyDeclaredDirsAndNgModules = [
+      ...this.localReader.getKnown(MetaKind.NgModule),
+      ...this.localReader.getKnown(MetaKind.Directive),
```
Why include directives here? AFAICS the `doesExport` method can only do two things for directives:

1. immediately return true if it equals `directiveOrPipe.ref.node`. This true value is ignored, so this is a no-op.
2. Register the directive reference in `references`, then fail to find NgModule metadata and return false. The record in `references` is never used though, as it's only recalled for NgModule classes, not directives.

I might be missing something here.

dylhunn replied:
I have included directives because, for standalone directives, I want to begin exporing from all of the directive's imports, so that we cover as much of the import/export graphs as possible. e.g. if a standalone directive imports `CommonModule`, I want to find `DatePipe`, even if it's not used in the program yet.

dylhunn commented on packages/compiler-cli/src/ngtsc/metadata/src/ng_module_index.ts:
```diff
@@ -0,0 +1,100 @@
+    // be reference-equal. However, I have a map on the side of every reference found to each
+    // ClassDeclaration, so at the end, I can look up *all* of the references for each result, and
+    // then ask for the corresponding NgModuleMetadata (which might be from a dts or local source).
+    // There might be a better way to handle this my adding new identifiers to references?
```
Specifically, I was contemplating whether calling `addIdentifier()` on some or all of the references could yield better code or a more efficient approach. What I'm doing works when tested, but may not be optimal.

"New" in this context refers to a `Reference<ClassDeclaration>` of an NgModule that was discovered while searching the `.imports` and `.exports` graphs.

dylhunn commented:
@JoostK Thanks for your review! I have applied all the suggested fixes, and left open a few points of discussion above.

alxhub commented on packages/compiler-cli/src/ngtsc/metadata/src/api.ts:
```diff
@@ -251,10 +251,10 @@ export interface MetadataReader {
  */
 export interface MetadataReaderWithIndex extends MetadataReader {
-  getKnown(kind: MetaKind): Iterable<ClassDeclaration>;
+  getKnown(kind: MetaKind): Array<Reference<ClassDeclaration>>;
```
I'm curious whether `Reference` is really necessary here.

A `Reference` is contextual - it wraps a declaration with information about how that declaration was identified in a particular context, such as an `import` array. Since the caller won't have any idea what context created that `Reference`, there's very little that they can reliably do with it.

An API that enumerates traits (e.g. all known directives) should probably return an `Array<ClassDeclaration>` instead. It's then up to the caller to look up an appropriate `Reference` given their use case.

alxhub commented on packages/compiler-cli/src/ngtsc/typecheck/testing/index.ts:
```diff
@@ -577,10 +577,9 @@ function getFakeMetadataReader(fakeMetadataRegistry: Map<any, DirectiveMeta|null
       switch (kind) {
-        case MetaKind.Directive:
-          return fakeMetadataRegistry.keys();
+        // TODO: This is not needed for these ngtsc tests, but may be wanted in the future.
```
Nit: should we throw instead of returning a false result here?

alxhub commented on packages/compiler-cli/src/ngtsc/metadata/src/api.ts:
```diff
@@ -257,6 +257,13 @@ export interface MetadataReaderWithIndex extends MetadataReader {
+ * An NgModuleIndex allows access to information about traits exported by NgModules.
+ */
+export interface NgModuleIndex {
+  getNgModulesExporting(trait: DirectiveMeta|PipeMeta): NgModuleMeta[];
```
Rather than deal in metadata, we should take and return `ClassDeclaration`s here.

dylhunn committed:
refactor(compiler-cli): Make `getKnown` return an array of nodes.

`MetadataReaderWithIndex.getKnown` currently returns an iterator. It will be easier to work with for upcoming usages if it returns an array instead.

dylhunn committed:
feat(language-service): Introduce a new NgModuleIndex, and use it to suggest re-exports.

NgModules can re-export other NgModules, which transitively includes all traits exported by the re-exported NgModule. We want to be able to suggest *all* re-exports of a component when trying to auto-import it.

Previously, we used an approximation when importing from NgModules: we looked at a Component's metadata, and just imported the declaring NgModule. However, this is not technically correct -- the declaring NgModule it is not necessarily the same one that exports it for the current scope. (Indeed, there could be multiple re-exports!) As a replacement, I have implemented a more general solution.

This PR introduces a new class on the template type checker, called `NgModuleIndex`. When queried, it conducts a depth-first-search over the NgModule import/export graph, in order to find all NgModules anywhere in the current dependency graph, as well as all exports of those NgModules. This allows the language service to suggest all of the re-exports, in addition to the original export.

dylhunn commented:
Back to @alxhub with a new design for `NgModuleIndex` -- we should be in the home stretch now!

alxhub commented on packages/compiler-cli/src/ngtsc/typecheck/src/checker.ts:
```diff
@@ -687,46 +688,57 @@ export class TemplateTypeCheckerImpl implements TemplateTypeChecker {
+  getPotentialImportsFor(
+      toImport: PotentialDirective|PotentialPipe,
+      inContext: ts.ClassDeclaration): ReadonlyArray<PotentialImport> {
+    const emit =
```
This code might be a little cleaner if this were a private method instead of a closure that we create here.

alxhub commented on packages/compiler-cli/src/ngtsc/typecheck/src/checker.ts:
```diff
@@ -687,46 +688,57 @@ export class TemplateTypeCheckerImpl implements TemplateTypeChecker {
+          const emitted = emittedRef.expression;
+          if (emitted instanceof WrappedNodeExpr) {
+            // An appropriate identifier is already in scope.
+            return {kind, symbolName: emitted.node.getText()};
```
Is `getText()` what we want here? If we _know_ that `emitted` is a `ts.Identifier` we should use `.text` instead. If not, then this doesn't seem safe.

alxhub commented on packages/compiler-cli/src/ngtsc/metadata/src/util.ts:
```diff
@@ -15,6 +15,8 @@ import {nodeDebugInfo} from '../../util/src/typescript';
 import {DirectiveMeta, DirectiveTypeCheckMeta, MetadataReader, NgModuleMeta, PipeMeta, TemplateGuardMeta} from './api';
 import {ClassPropertyMapping, ClassPropertyName} from './property_mapping';
 
+export const nonNull = <T>(m: T|null): m is T => m !== null;
```
This doesn't appear to be used.

dylhunn commented:
caretaker: the failing windows CI job appears to be unrelated.

AndrewKushnir commented:
This PR was merged into the repository by commit 4ae384fd619a13eaadf737d08a97f07e1f6b273c.
